src/Hyperion/Benchmark.hs

```diff
@@ -55,7 +60,37 @@ instance Show Benchmark where
   showsPrec x (WithSampling _opt bk) = showsPrec x bk
 
 bench :: String -> Batch () -> Benchmark
-bench name batch = Bench (Text.pack name) batch
+bench name batch = Bench Text.pack (const []) id (const batch) name
+
+benchWith
+  :: (a -> Text)
+  -> (a -> [JSON.Pair])
+  -> (a -> b)
+  -> (b -> Batch ())
+  -> a -> Benchmark
+benchWith = Bench
+
+benchMany
```

Comment:
> I don't understand the use case for benchMany. It looks to me like you're reinventing an existing functionality in Hyperion, namely "series". Could you clarify how the two differ?

`series` does output some of the user metadata in the `bench_name`. However we can't expect the user to parse it and go fish out the input element's `show` output.

`benchWith` uses the new `Bench` constructor for recording the metadata. I could have added it to the `Series` constructor instead (see internal discussion about that) but adding it to `Bench` means that single benchmarks may have extra metadata as well.

> The idea with series is precisely to support one of the use cases you mention in your description

Take the following, from `examples/Main.hs`:

``` haskell
  series [10..10] $ \n ->
    bgroup "pure-functions"
      [ bench "fact" (use n >>= nf fact)
      , bench "fib" (use n >>= nf fib)
      ]
```

How could I plot the durations of `pure-functions/n:fact` against `n` if I don't have `n` to start with? Once again, while it's theoretically possible to parse the name and get the parameter value out, it's really not convenient.